{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "365d5c8d",
   "metadata": {},
   "source": [
    "## **Raw FASTA file for H37Rv reference genome**\n",
    "Used for genome assembly, alignments and sequence search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cd9573cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = \"raw/GCF_000195955.2_ASM19595v2_genomic.fna.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3c9e2337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_id</th>\n",
       "      <th>description</th>\n",
       "      <th>sequence</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000962.3</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>TTGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACGCGG...</td>\n",
       "      <td>4411532</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sequence_id                                        description  \\\n",
       "0  NC_000962.3  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "\n",
       "                                            sequence   length  \n",
       "0  TTGACCGATGACCCCGGTTCAGGCTTCACCACAGTGTGGAACGCGG...  4411532  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Raw Fasta\n",
    "\n",
    "import gzip\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "\n",
    "def fasta_to_dataframe(fasta_gz_path):\n",
    "    # Read the compressed FASTA file\n",
    "    with gzip.open(fasta_gz_path, 'rt') as f:\n",
    "        fasta_data = f.read()\n",
    "    \n",
    "    # Parse using Biopython's SeqIO\n",
    "    records = SeqIO.parse(StringIO(fasta_data), 'fasta')\n",
    "    \n",
    "    # Create DataFrame\n",
    "    data = []\n",
    "    for record in records:\n",
    "        data.append({\n",
    "            'sequence_id': record.id,\n",
    "            'description': record.description,\n",
    "            'sequence': str(record.seq),\n",
    "            'length': len(record.seq)\n",
    "        })\n",
    "    \n",
    "    return pd.DataFrame(data)\n",
    "\n",
    "df = fasta_to_dataframe(fasta_file)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "46305093",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_parquet('staging/reference_linear_genome.parquet', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c0317f3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>genome_type</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>5</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id      genome_type                                         chromosome  \\\n",
       "0   1  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "1   2  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "2   3  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "3   4  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "4   5  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "\n",
       "   position base  \n",
       "0         1    T  \n",
       "1         2    T  \n",
       "2         3    G  \n",
       "3         4    A  \n",
       "4         5    C  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Formatted Indexed Fasta\n",
    "import pandas as pd\n",
    "\n",
    "def fasta_to_positional_parquet(fasta_gz_path, output_parquet_path, genome_type=\"H37Rv reference\"):\n",
    "    # Read the FASTA file into DataFrame (using your existing function)\n",
    "    df_fasta = fasta_to_dataframe(fasta_gz_path)\n",
    "    \n",
    "    # Prepare an empty list to store all positional records\n",
    "    positional_data = []\n",
    "    \n",
    "    # Process each sequence in the FASTA DataFrame\n",
    "    for _, row in df_fasta.iterrows():\n",
    "        sequence_id = row['sequence_id']\n",
    "        description = row['description']\n",
    "        sequence = row['sequence']\n",
    "        \n",
    "        # Create a record for each base in the sequence\n",
    "        for position, base in enumerate(sequence, start=1):\n",
    "            positional_data.append({\n",
    "                'genome_type': genome_type,\n",
    "                'chromosome': description,\n",
    "                'position': position,\n",
    "                'base': base\n",
    "            })\n",
    "    \n",
    "    # Create DataFrame from the collected records\n",
    "    df_positional = pd.DataFrame(positional_data)\n",
    "    \n",
    "    # Add an auto-incrementing ID (similar to SQLite)\n",
    "    df_positional.reset_index(inplace=True)\n",
    "    df_positional.rename(columns={'index': 'id'}, inplace=True)\n",
    "    df_positional['id'] += 1  # Start IDs at 1 instead of 0\n",
    "    \n",
    "    # Save to Parquet file\n",
    "    df_positional.to_parquet(output_parquet_path, index=False)\n",
    "    \n",
    "    return df_positional\n",
    "\n",
    "# Usage example\n",
    "output_file = \"staging/reference_genome.parquet\"\n",
    "result_df = fasta_to_positional_parquet(fasta_file, output_file)\n",
    "\n",
    "result_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f7c06557",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>genome_type</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>2</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>5</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>6</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>7</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>8</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>9</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>H37Rv reference</td>\n",
       "      <td>NC_000962.3 Mycobacterium tuberculosis H37Rv, ...</td>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id      genome_type                                         chromosome  \\\n",
       "0   1  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "1   2  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "2   3  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "3   4  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "4   5  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "5   6  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "6   7  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "7   8  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "8   9  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "9  10  H37Rv reference  NC_000962.3 Mycobacterium tuberculosis H37Rv, ...   \n",
       "\n",
       "   position base  \n",
       "0         1    T  \n",
       "1         2    T  \n",
       "2         3    G  \n",
       "3         4    A  \n",
       "4         5    C  \n",
       "5         6    C  \n",
       "6         7    G  \n",
       "7         8    A  \n",
       "8         9    T  \n",
       "9        10    G  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Database integration Validation\n",
    "\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "\n",
    "# Connect to the database\n",
    "conn = sqlite3.connect('mtb_warehouse.db')\n",
    "\n",
    "# Method 1: Using read_sql() - recommended approach\n",
    "query = \"SELECT * FROM reference_genome LIMIT 10;\"\n",
    "df = pd.read_sql(query, conn)\n",
    "\n",
    "# Method 2: Using fetchall() and converting to DataFrame\n",
    "cursor = conn.cursor()\n",
    "cursor.execute(\"SELECT * FROM reference_genome LIMIT 10;\")\n",
    "rows = cursor.fetchall()\n",
    "column_names = [description[0] for description in cursor.description]  # Get column names\n",
    "df2 = pd.DataFrame(rows, columns=column_names)\n",
    "\n",
    "# Close the connection\n",
    "conn.close()\n",
    "\n",
    "# Display the DataFrame\n",
    "display(df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
